### PR TITLE
Async comp builder

### PR DESCRIPTION
**Pull request by drakos74**
## Which problem is this PR solving?

Async Component Builder Pattern #54

## Short description of the changes

- introduced aync component builder
(following the suggestion in the issue, backwards compatibility has been broken)
p.s. If this is not acceptable, we can keep the 'New' method as the component constructor, and use another name for the builder constructor
- introduced iota for enum of Failure Strategies and a map for values ,
to make comparisons regarding this enum easier
- extended test scope to cover ALL failure scenarios / error code paths
- refactor tests for async component, to make them more compact and readable
- re-used string templates for repeating log or error messages
- adjusted the existing examples to use the builder

### CODE REVIEW

File: async/component.go [hunk 1]
```diff
@@ -38,36 +41,102 @@ type Component struct {
...
-		proc:         p,
-		cf:           cf,
-		failStrategy: NackExitStrategy,
-		retries:      0,
-		retryWait:    0,
+// WithFailureStrategy defines the failure strategy to be used
+// default value is NackExitStrategy
+// it will append an error to the builder if the strategy is not one of the pre-defined ones
+func (cb *Builder) WithFailureStrategy(fs FailStrategy) *Builder {
+	if _, ok := failStrategyValues[fs]; ok {
```
Comment (mantzas) [inline async/component.go]: As I understand it, we are using the map to check if we support the failurestrategy or if and invalid value has been passed in.
If that is the reason we should do a simple switch statement to see if the value is acceptable or not, and avoid having an map allocated

**Review by mantzas — CHANGES_REQUESTED** [review]

File: async/component.go [hunk 2]
```diff
@@ -38,36 +41,102 @@ type Component struct {
...
+		name:   name,
+		errors: errs,
+	}
+}
 
-	if p == nil {
-		return nil, errors.New("work processor is required")
+// WithProcessor sets the processing logic for the component
+// it will append an error to the builder if the processor is nil
+func (cb *Builder) WithProcessor(proc ProcessorFunc) *Builder {
```
[the remark this answers is not in the document]
Comment (drakos74) [reply async/component.go]: agreed , just followed the example in the issue. 
Will adjust

File: async/component.go [hunk 3]
```diff
@@ -2,13 +2,16 @@ package async
...
 	"context"
+	"fmt"
 	"time"
 
 	"github.com/beatlabs/patron/errors"
 	"github.com/beatlabs/patron/log"
 	"github.com/prometheus/client_golang/prometheus"
 )
 
+var propSetMSG = "property '%s' set for '%s'"
```
[the remark this answers is not in the document]
Comment (drakos74) [reply async/component.go]: yeap!

File: async/component.go [hunk 4]
```diff
@@ -148,6 +217,9 @@ func (c *Component) processMessage(msg Message, ch chan error) {
 	}
 }
 
+var errInvalidFS = errors.New("invalid failure strategy")
+var failureStrategyErrorMSG = "%s failed when executing failure strategy"
```
[the remark this answers is not in the document]
Comment (drakos74) [reply async/component.go]: yeap!

**Review by mantzas — DISMISSED** [review]

**Review by moukoublen — DISMISSED** [review]

**Review by moukoublen — APPROVED** [review]

**Review by mantzas — APPROVED** [review]

**Pull request merged**